### PR TITLE
Update harnesses for aes_gcm_encrypt and aes_gcm_decrypt

### PR DESCRIPTION
**Pull request by tegansb**
*Description of changes:*
Added assumptions that the output buffer cannot be NULL. 
This assumption is needed. If the output buffer is NULL, the EVP_EncryptUpdate or EVP_DecryptUpdate stubs assume the case where the aad is being specified and do not update outl. This results in an overflow error as there are no constraints on the size of outl. The behavior of the stubs does seem to match the specification for the functions so I updated the assumptions instead of the stubs. 
Added matching preconditions to source code.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

# Check any applicable:
- [ ] Were any files moved? Moving files changes their URL, which breaks all hyperlinks to the files.

### CODE REVIEW

**Review by alex-chew — APPROVED** [review]
Overall LGTM. Do we have a mechanism to ensure that the preconditions in each operation under verification correspond to an assumption in the harness, or are we just doing that manually for now?

File: source/cipher.c [hunk 1]
```diff
@@ -814,6 +814,13 @@ int aws_cryptosdk_aes_gcm_encrypt(
     const struct aws_byte_cursor iv,
     const struct aws_byte_cursor aad,
     const struct aws_string *key) {
+    AWS_PRECONDITION(aws_byte_buf_is_valid(cipher));
+    AWS_PRECONDITION(cipher->buffer != NULL);
+    AWS_PRECONDITION(aws_byte_buf_is_valid(tag));
+    AWS_PRECONDITION(aws_byte_cursor_is_valid(&plain));
+    AWS_PRECONDITION(aws_byte_cursor_is_valid(&iv));
+    AWS_PRECONDITION(aws_byte_cursor_is_valid(&aad));
+    AWS_PRECONDITION(key != NULL);
```
Comment (feliperodri) [inline source/cipher.c:823]: Why not `AWS_PRECONDITION(aws_string_is_valid(key));`?

File: source/cipher.c [hunk 2]
```diff
@@ -854,6 +861,13 @@ int aws_cryptosdk_aes_gcm_decrypt(
     const struct aws_byte_cursor iv,
     const struct aws_byte_cursor aad,
     const struct aws_string *key) {
+    AWS_PRECONDITION(aws_byte_buf_is_valid(plain));
+    AWS_PRECONDITION(plain->buffer != NULL);
+    AWS_PRECONDITION(aws_byte_cursor_is_valid(&cipher));
+    AWS_PRECONDITION(aws_byte_cursor_is_valid(&tag));
+    AWS_PRECONDITION(aws_byte_cursor_is_valid(&iv));
+    AWS_PRECONDITION(aws_byte_cursor_is_valid(&aad));
+    AWS_PRECONDITION(key != NULL);
```
Comment (feliperodri) [inline source/cipher.c:870]: Why not `AWS_PRECONDITION(aws_string_is_valid(key));`?

**Review by feliperodri — CHANGES_REQUESTED** [review]
I have a couple questions. I also recommend to update the AWS C Common submodule to the latest version. @tegansb

File: source/cipher.c (hunk 1 above)
Comment (tegansb) [reply source/cipher.c]: True that's a better check! Updated and also updated to latest c-common

File: source/cipher.c (hunk 2 above)
Comment (tegansb) [reply source/cipher.c]: Same as above. Updated!

**Review by feliperodri — APPROVED** [review]
LGTM.

**Pull request merged**